src/main/scala/Ch4.scala

```diff
@@ -0,0 +1,204 @@
+
+  // Sub-structure is covariant wrt B since it's not consumed
+  type BOrXToB[B, X] = Either[B, X => B]
+
+  def fmapBOrXToB[B, C, X](f: B => C): BOrXToB[B, X] => BOrXToB[C, X] = {
+    case Left(b) => Left(f(b))
+    case Right(xToB) => Right(x => f(xToB(x)))
+  }
+
+  def p5Fmap[B, C](f: B => C): P5Data[B] => P5Data[C] = { p5DB =>
+    val (bOrIntToB, bOrStrToB) = p5DB
+    (fmapBOrXToB(f)(bOrIntToB), fmapBOrXToB(f)(bOrStrToB))
+  }
+
+
+  // Problem 6
+  // Short notation:
+  // Data[A, B] ≡ ((A x B) x (B => Int)) + (A x B x Int) + ((String => A) x (B => A))
+  // A is never consumed, so it is covariant
+  // B behaves like a functor in case class `Re` but is consumed in the others,
+  // so it's invariant
+  sealed trait Coi[+A, B]
+  case class Pa[+A, B](b: (A, B), c: B⇒Int) extends Coi[A, B]
+  case class Re[+A, B](d: A, e: B, c: Int) extends Coi[A, B]
+  case class Ci[+A, B](f: String⇒A, g: B⇒A) extends Coi[A, B]
+
+}
+
+
+object Ch4Ex2 {
+
+  /**
+    Problem 1
+
+    Given functors F[A] and G[A] prove F[A] x G[A] is also a functor
+    We know following functions exist: fmapF(f): F[A] => F[B] and fmapG(f): G[A] => G[B]
+
+    Define fmap<FxG>[A, B](f: A => B) = (p: F[A], q: G[A]) => (fmapF(f)(p), fmapG(f)(q))
+
+    Check identity
+    fmap(id[A])((p, q)) = (fmapF(id[A])(p), fmapG(id[A])(q)) = (p, q)
+
+    Check composition
+    (fmap<FxG>(f1) ◦ fmap<FxG>(f2))(p, q)
+    = fmap(f2)(fmapF(f1)(p), fmapG(f1)(q))
+    = ((fmapF(f1) ◦ fmapF(f2))(p), (fmapG(f1) ◦ fmapG(f2))(q)
+    = (fmapF(f1 ◦ f2)(p), fmapG(f1 ◦ f2)(q))
+    = fmap(f1 ◦ f2)(p, q)
+    */
+
+  // Problem 2
+  // type FToG[A] = F[A] => G[A]
+  // // functor
+  // type F[A] = Option[A]
+  // // functor
+  // type G[A] = (A, Int)
+
+  // def fmapF[A, B](f: A => B): Option[A] => Option[B] = {
+  //   case Some(a) => Some(f(a))
+  //   case None => None
+  // }
+
+  // def fmapG[A, B](f: A => B): (A, Int) => (B, Int) = {
+  //   case (a, i) => (f(a), i)
+  // }
+
+  // Cannot implement fmap
+  // def fmapFToG[A, B](f: A => B): FToG[A] => FToG[B] = { fToGA =>
+  //   val fToGB: FToG[B] = (fB: F[B]) => {
+  //     val bToA: B => A = ??? // Need a function that maps B to A
+  //     val fA: F[A] = fmapF[B, A](bToA)(fB) // or we need contraFmapF[B, A](f)(fB)
+  //     val gA: G[A] = fToGA(fA)
+  //     val gB: G[B] = fmapG(f)(gA)
+  //     gB
+  //   }
+  //   fToGB
+  // }
+
+  // Cannot implement contraFmap
+  // def contraFmapFToG[A, B](f: B => A): FToG[A] => FToG[B] = { fToGA =>
+  //   val fToGB: FToG[B] = (fB: F[B]) => {
+  //     val fA: F[A] = fmapF[B, A](f)(fB)
+  //     val gA: G[A] = fToGA(fA)
+  //     val aToB: A => B = a => ???  // There is no correct way to produce a value of type B
+  //     val gB: G[B] = fmapG(aToB)(gA)
+  //     gB
+  //   }
+  //   fToGB
+  // }
+
+  // Since we cannot implement fmap for FToG[A], it is not a functor
+  // and since we cannot implement contrafmap, it is not a contrafunctor
```

Comment:
This is correct. You would have gotten the same result a little quicker if you chose F[A] = A and G[A] = A (the identity functor).

Reply:
lol that's true